PhysicsTools/TensorFlow/src/TensorFlow.cc

```diff
@@ -25,6 +25,23 @@ namespace tensorflow {
     setThreading(sessionOptions, nThreads);
   }
 
+  void setBackend(SessionOptions& sessionOptions, Backend backend) {
+    if (backend == Backend::cpu) {
+      // hotfix: disable GPU usage whatsoever for now, add a convenient interface in a future PR
+      (*sessionOptions.config.mutable_device_count())["GPU"] = 0;
+      sessionOptions.config.mutable_gpu_options()->set_visible_device_list("");
+    } else if (backend == Backend::cuda) {
+      // Options from https://github.com/tensorflow/tensorflow/blob/c53dab9fbc9de4ea8b1df59041a5ffd3987328c3/tensorflow/core/protobuf/config.proto#L20
+      edm::LogInfo("PhysicsTools/TensorFlow") << "Setting up TensorFlow to use 1 GPU";
+      edm::LogInfo("PhysicsTools/TensorFlow")
+          << "Visible GPU devices << " << sessionOptions.config.mutable_gpu_options()->visible_device_list();
+      // Take only the first GPU in the CUDA_VISIBLE_DEVICE list
+      (*sessionOptions.config.mutable_device_count())["GPU"] = 1;
```

Comment:
How would this setting look like for AMD (or Intel) GPUs? Would it be the same ? (I'd guess so)

In case the code would be the same, I'm wondering if there is still benefit in separating the vendors by `Backend::cuda`, `Backend::rocm`, `Backend::<intel>`, or would it make sense to simplify with `Backend::gpu`? (I'm a little bit leaning towards keeping the separation, but wanted to ask anyway)

Reply:
Since you mention that... what happens when both NVIDIA (CUDA) and AMD (ROCm) GPUs are available ?

Reply:
At the moment the tensorflow build works only for Nvidia GPU. I'm adding  a check with `HeterogeneousCore/CUDAUtilities` and `HeterogeneousCore/ROCmUtilities` to make the code aware of the actual presence of a device.  By default the code is taking the first device in the list of available devices (from CUDA_VISIBLE_DEVICES).  May I ask you some additional info about how the list of devices is propagated to the different CMSSW processes when there are more than one? 

I would make it work as:
- If the user wants `Backend::cuda` and a cuda device is present, enable it and take the first device available in the list.
- If the user wants `Backend::cuda`, but no cuda device is present, or a ROCm device is present --> fallback on CPU
- If the user wants `Backend::rocm` --> fallback on CPU (for the moment)

Doing so we can be sure about the matching between the user request and the present devices. On the other hand this mechanism won't automatically activate the GPU for producers using TensorFlow when a device is present. I think this type of automatic selection mechanism is something still under discussion , right?

Reply:
> At the moment the tensorflow build works only for Nvidia GPU

Do you know if a single TensorFlow build can use both NVIDIA GPUs and AMD GPUs? Or would we need two different builds?

> I'm adding a check with `HeterogeneousCore/CUDAUtilities` and `HeterogeneousCore/ROCmUtilities` to make the code aware of the actual presence of a device.

I would advise against that. We want to be able to have CMSSW builds without CUDA and without ROCm. If you add direct dependencies to CUDA or ROCm here, they would need to be protected with `#ifdef`s, which could end up in a complicated setup. Instead, I would now look into utilizing the `edm::ResourceInformation` service that is independent of CUDA/ROCm/etc, but contains information if any such devices are present (we could think of adding the number of CUDA devices and ROCm devices there as well if that would be useful, see also https://github.com/cms-sw/cmssw/pull/39402/files#r1093351644).

> By default the code is taking the first device in the list of available devices (from CUDA_VISIBLE_DEVICES). May I ask you some additional info about how the list of devices is propagated to the different CMSSW processes when there are more than one?

CMSSW follows the `CUDA_VISIBLE_DEVICES` (and `HIP_VISIBLE_DEVICES`). Internally the CUDA (to be phased out) and Alpaka framework parts decide for each EDM stream what device the CUDA/Alpaka EDModules on that stream should use. I think we should think more about the desired behavior of a job with ML inference and Alpaka/CUDA modules on a multi-GPU setup for longer term (i.e. after this PR, maybe also in an ML group or a core software meeting?), and now go with something simple and reasonable choice (I think for now just using first device would be sufficient).

> * If the user wants `Backend::cuda` and a cuda device is present, enable it and take the first device available in the list.
> * If the user wants `Backend::cuda`, but no cuda device is present, or a ROCm device is present --> fallback on CPU
> * If the user wants `Backend::rocm` --> fallback on CPU (for the moment)

I think if the user explicitly asks for `Backend::cuda`, the job should terminate with an exception, and the behavior of "use GPU if possible, otherwise fall back to CPU" would be a separate explicit option. We want to be able to catch early problems where a job is explicitly told to use a GPU, but something goes wrong in the setup and the job does not see the GPU. In such a case it can be better to abort the job and let the operations know about a problem rather than making the jobs slow. E.g. the `process.options.accelerators` and the ONNX PR https://github.com/cms-sw/cmssw/pull/39402 follow this principle.

> I think this type of automatic selection mechanism is something still under discussion , right?

In case of ONNX the decision was to make the backend configurable for each EDModule (in https://github.com/cms-sw/cmssw/pull/36963). The configuration parameter could be used with `SwitchProducer` for an automatic selection mechanism. But I think we should be able to achieve a solution with less boilerplate. I also think we should strive to have similar interface and behavior for both ONNX and TensorFlow (modules) in this respect.

Reply:
Thanks a lot @makortel for all the pointer.  I think I understood what has to be done. I will commit an implementation soon.

Reply:
> Do you know if a single TensorFlow build can use both NVIDIA GPUs and AMD GPUs? Or would we need two different builds?

The current TensorfFlow build works only for NVIDIA GPUs.  AMD would need a different build, we haven't looked into it yet.

Reply:
> > Do you know if a single TensorFlow build can use both NVIDIA GPUs and AMD GPUs? Or would we need two different builds?
> 
> The current TensorfFlow build works only for NVIDIA GPUs. AMD would need a different build, we haven't looked into it yet.

Thanks. I understand adding AMD GPU support would require a new TensorFlow build. I'm specifically interested if, in principle, one TensorFlow build can support NVIDIA and AMD GPUs, or if we'd need one TensorFlow build for NVIDIA and another build for AMD? (latter leading to some complexity, which is why it would be good to know early)

Reply:
i think there will be two tensorflow builds. there are specific tensorflow versions for amd gpus (called tensorflow-rocm) https://rocmdocs.amd.com/en/latest/Deep_learning/Deep-learning.html#tensorflow

Reply:
Hi @makortel,  I think I have implemented your suggestion.  The only thing is that when I run locally the CUDA variant tests I get an error because no `ServiceRegistry` can be found. 

```

##Failure Location unknown## : Error
Test name: testSessionCacheCUDA::test
uncaught exception of type std::exception (or derived).
- An exception of category 'NotFound' occurred.
Exception Message:
No ServiceRegistry has been set for this thread. Tried to get a Service edm::ResourceInformation
```

How can I solve this? 

#### ONNX and TF interface 
> In case of ONNX the decision was to make the backend configurable for each EDModule (in https://github.com/cms-sw/cmssw/pull/36963). The configuration parameter could be used with SwitchProducer for an automatic selection mechanism. But I think we should be able to achieve a solution with less boilerplate. I also think we should strive to have similar interface and behavior for both ONNX and TensorFlow (modules) in this respect.

I completely agree we should have the goal to unify the behaviour of the TF and ONNX interface. I will contact @davidlange6  for https://github.com/cms-sw/cmssw/pull/39402. I think we can create a follow up PR unifying the interfaces, and adding fallbacks for ROCm also.

Reply:
> How can I solve this?

You can find an example of setting up the Service system (with `ResourceInformationService`) at the end of https://github.com/cms-sw/cmssw/pull/40625 . I find it somewhat annoying that the testing becomes more difficult (i.e. we add more coupling here), but maybe that is ok. Note that for CUDA tests you need to also include `process.add_(cms.Service('CUDAService'))` in the configuration fragment, or `edm::ResourceInformation` looks like there would be no GPUs.

Reply:
Ok thanks!  I think that I fixed this in the latest commit.  Tests are running locally for me. 

I'm still testing on an environment without the TF-GPU build ON, and the result of the tests on a machine with Nvidia GPu is: 

- CUDA service is active
- the TF backend==cuda looks for the CUDA service via `ResourceInformation`, finds a CUDA device, and does not complain. 
- TF does not see any device though (because I'm running on a standard build), but it does not complain and fall back on CPU. 

Maybe should I add an explicit check that TF is going to actually see and use the device? and otherwise raise an expection?

Reply:
> * TF does not see any device though (because I'm running on a standard build), but it does not complain and fall back on CPU.

Just to check that I understood, in absence of a GPU, TF itself falls back to CPU, even if it was instructed to use GPU?

> Maybe should I add an explicit check that TF is going to actually see and use the device? and otherwise raise an exception?

I think it would be a good idea. At least the behavior would become consistent with the "if GPU is asked without fallback, the job aborts if it is unable to use the GPU". How difficult it is to check what TF will actually see?

Reply:
> Just to check that I understood, in absence of a GPU, TF itself falls back to CPU, even if it was instructed to use GPU?
That's correct. When TF starts to assign operators to devices it just uses the CPU since it doesn't see any GPU. 

> I think it would be a good idea. At least the behavior would become consistent with the "if GPU is asked without fallback, the job aborts if it is unable to use the GPU". How difficult it is to check what TF will actually see?

I think it's feasible. I'm investigating and I will update the PR soon.

Reply:
I have added a couple of unittest on the `visible_device` API in Tensorflow. 

In principle now if the `ResourceInformation` service confirms the presence of a CUDA device, and the user has requested the `Backend::cuda`, the device is used for sure.   TF was falling back on CPU silently only because the test was run without the TF-gpu build active. 

One of the unit test `testTFVisibleDeviceCUDA` will fail when this PR is  tested without https://github.com/cms-sw/cmsdist/pull/7648: that's because the CUDA device is not seen by TF.     How should we handle this? 

Thanks.

Reply:
> One of the unit test `testTFVisibleDeviceCUDA` will fail when this PR is tested without [cms-sw/cmsdist#7648](https://github.com/cms-sw/cmsdist/pull/7648): that's because the CUDA device is not seen by TF. How should we handle this?

This PR and https://github.com/cms-sw/cmsdist/pull/7648 just need to be tested and merged together (which you already started on the cmsdist side). Or did you have some other concern in mind?

Reply:
No no ok, it was just that. I see that tests are already starting, thanks again for the support!